### PR TITLE
Swift 3 compatibility

### PR DESCRIPTION
**Pull request by mbrandonw**
This updates KsApi to be Swift 3 compatible. This does not change our APIs to be more "Swift 3 like". We should do that in a separate PR so that we can concentrate only on Swift 3 and not a bunch of other API-breaking changes.

This PR depends on the following other PR's:

https://github.com/kickstarter/Kickstarter-Prelude/pull/62
https://github.com/kickstarter/Kickstarter-ReactiveExtensions/pull/54

#### Notes

This migration was way bigger and more subtle than the other two. Some things to look out for:

* We had to upgrade Argo to [4.1.0](https://github.com/thoughtbot/Argo/releases/tag/v4.1.0), which came with a few changes from our current version:
  * Argo now requires [Runes](https://github.com/thoughtbot/Runes) as an explicit dependency (used to just be a submodule managed by Carthage). This just means we gotta `import Runes`anytime we do Argo stuff.
  * The semantics of decoding an optional has slightly changed https://github.com/thoughtbot/Argo/pull/420. It's for the better, but it's subtle and caused a few things to break in our tests. It all stems from the decision of how to handle decoding an optional value that fails. Should that fail the whole decoding, or should it be lifted to a `.success(nil)`?
* A _ton_ of `AnyObject`s have changed to `Any`s. I'm not entirely up-to-date with what changed in Swift to require this, but it's def for the better.
* I've got more notes, but I'll do em inline...

### CODE REVIEW

File: KsApi/lib/MimeType.swift [hunk 1]
```diff
@@ -1,10 +1,13 @@
 import Foundation
 import MobileCoreServices
 
-extension NSData {
+extension Data {
   internal var imageMime: String? {
-    guard let byte: UInt8 = UnsafeBufferPointer(start: UnsafePointer(self.bytes), count: 1).first
-      else { return nil }
+
+    let start = (self as NSData).bytes.bindMemory(to: UInt8.self, capacity: self.count)
```
Comment (mbrandonw) [inline KsApi/lib/MimeType.swift]: @stephencelis the migrator did this and i really don't know what it's doing. if i try to get rid of the `NSData` cast it suggests to use `withUnsafeBytes`, but i'm not quite able to get the whole thing going...

File: KsApi/models/User.swift [hunk 2]
```diff
@@ -67,14 +68,14 @@ extension User: CustomDebugStringConvertible {
...
-  public static func decode(json: JSON) -> Decoded<User> {
+  public static func decode(_ json: JSON) -> Decoded<User> {
     let create = curry(User.init)
     let tmp = create
       <^> json <| "avatar"
       <*> json <|? "facebook_connected"
       <*> json <| "id"
       <*> json <|? "is_friend"
-      <*> json <|? "location"
+      <*> (json <|? "location" <|> .success(nil))
```
Comment (theginbin) [inline KsApi/models/User.swift]: why did you coalesce to `.success` here? it's optional...

File: KsApiTests/lib/EncodableTests.swift [hunk 3]
```diff
@@ -19,7 +22,7 @@ final class EncodableTests: XCTestCase {
   func testToJSONData() {
     let model = EncodableModel(id: 1, name: "Blob")
     let jsonString = "{\"ID\":1,\"NAME\":\"Blob\"}"
-    let jsonData = jsonString.dataUsingEncoding(NSUTF8StringEncoding)
+    let jsonData = jsonString.data(using: String.Encoding.utf8)
```
Comment (theginbin) [inline KsApiTests/lib/EncodableTests.swift]: does just `.utf8` work here?

Comment (mbrandonw) [inline KsApiTests/lib/EncodableTests.swift]: it does!

File: KsApi/models/User.swift (hunk 2 above)
Comment (mbrandonw) [inline KsApi/models/User.swift]: this is the subtle argo change that i mentioned in the description of this PR. check it out and lemme know if you have q's.

**Comment by mbrandonw** [comment]
Updated all the dependencies and a lil more spot checking but this this is g2g once the other stuff is!

**Pull request merged**